### ISSUE TITLE
HTTP3 + QUIC support

### ISSUE THREAD

**Issue by nyanmark** [jc21/nginx-proxy-manager#1550]
With QUIC support being in most of today's browsers, and tools already excising to enable QUIC within nginx I believe it would be a good idea to add the support to NPM helping users speed up those huge video streams.

**Comment by iamgadgetman** [comment]
It would be nice to be able to use NPM to connect to internet radio servers that use QUIC.

**Comment by ingarb** [comment]
Hope this can be added to NPM. I would imagine if you don't have QUIC Google might rank your site lower copared to a site that does have QUIC. It would also make your site a little fit faster - also leading to a higher ranking.

**Comment by danfratamico** [comment]
I totally agree here, this needs to be a top priority.

**Comment by timnolte** [comment]
I was going about the route of using NPM in front of a Docker/container setup only to realize that although the web server I'm running should support QUIC & HTTP/3 unfortunately NPM doesn't. Having NPM in front of any containers/services I start up is very handy but I'd sure like the ability to use QUIC & HTTP/3. Keeping an eye on this.

**Comment by nonlinearsugar** [comment]
It's still experimental/beta in Nginx and only available as separate packages/in a separate branch. https://www.nginx.com/blog/binary-packages-for-preview-nginx-quic-http3-implementation/ https://github.com/nginx-quic/nginx-quic

While I lend my support to those asking, I don't think this can/should go anywhere on the NPM side before Nginx merge it into their stable branch.

**Comment by EnfermeraSexy** [comment]
> It's still experimental/beta in Nginx and only available as separate packages/in a separate branch. https://www.nginx.com/blog/binary-packages-for-preview-nginx-quic-http3-implementation/ https://github.com/nginx-quic/nginx-quic
> 
> While I lend my support to those asking, I don't think this can/should go anywhere on the NPM side before Nginx merge it into their stable branch.

It's already merged, right?

**Comment by nonlinearsugar** [comment]
Yeah.  Time permitting, I think this is ready for @jc21 to action.

https://nginx.org/en/docs/quic.html

**Comment by squadramunter** [comment]
Any update on this?

**Comment by multipliedtwice** [comment]
Would be amazing to have it

**Comment by stanthewizzard** [comment]
would be awesome ... no news ?

**Comment by LZNetwork-Main** [comment]
I hope we will see this added in the near future, I am already using NGINX with HTTP/3 on some of my servers and it can improve things significantly with larger sites!
Please, this needs to be implemented. As the comment above said, it is already in mainline version of NGINX.

**Comment by devius** [comment]
so please, me too....
I am waiting for it.....

**Comment by LDprg** [comment]
Also all big Websites now support http/3 (Google, Microsoft, Cloudflare, Amazon, Meta). So I think it is getting time for a implementation.

**Comment by FluffyDiscord** [comment]
Any updates?

**Comment by fengyuclub** [comment]
Any updates?

**Comment by iwyz0** [comment]
Any updates?

**Comment by scampower3** [comment]
OpenResty 1.25.3.1 added support for HTTP/3.

**Comment by lateautumn233** [comment]
Hello, I have tried to add HTP3 support for nginx-proxy-manager

If you are interested in this, you can give it a try

docker: https://hub.docker.com/repository/docker/lateautumn233/nginx-proxy-manager/
Source code: https://github.com/lateautumn233/nginx-proxy-manager

**Comment by iwyz0** [comment]
> Hello, I have tried to add HTP3 support for nginx-proxy-manager
> 
> If you are interested in this, you can give it a try
> 
> docker: https://hub.docker.com/repository/docker/lateautumn233/nginx-proxy-manager/ Source code: https://github.com/lateautumn233/nginx-proxy-manager

You did what they couldn't do. Thank you

**Comment by daedric7** [comment]
Should this be as simple as replacing docker image in the compose?

**Comment by lateautumn233** [comment]
Yes
But currently only the amd64 architecture is available












lateautumn233
***@***.***


---Original---
From: ***@***.***&gt;
Date: Sat, Jan 27, 2024 01:21 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [NginxProxyManager/nginx-proxy-manager] HTTP3 + QUIC support(Issue #1550)




 
Should this be as simple as replacing docker image in the compose?

**Comment by daedric7** [comment]
> > Yes But currently only the amd64 architecture is available lateautumn233 _**@**_.***
> > […](#)
> > ---Original--- From: _**@**_._**> Date: Sat, Jan 27, 2024 01:21 AM To: _**@**_.**_>; Cc: _**@**_._**_**@**_.**_>; Subject: Re: [NginxProxyManager/nginx-proxy-manager] HTTP3 + QUIC support(Issue #1550) Should this be as simple as replacing docker image in the compose? — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented.Message ID: _**@**_.***>
> 
> I did it.
> 
> It seems H3 is being enabled:
> 
> ```
> npm-1  | [1/26/2024] [5:28:15 PM] [Migrate  ] › ℹ  info      Current database version: none
> npm-1  | [1/26/2024] [5:28:15 PM] [Migrate  ] › ℹ  info      [http3_support] Migrating Up...
> npm-1  | [1/26/2024] [5:28:15 PM] [Migrate  ] › ℹ  info      [http3_support] proxy_host Table altered
> npm-1  | [1/26/2024] [5:28:15 PM] [Migrate  ] › ℹ  info      [http3_support] redirection_host Table altered
> npm-1  | [1/26/2024] [5:28:15 PM] [Migrate  ] › ℹ  info      [http3_support] dead_host Table altered
> ```
> 
> But, in chrome, only H2 is being used:
> 
> ![image](https://private-user-images.githubusercontent.com/4562461/300060030-d24bbba8-743c-487a-84fc-628c74918218.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYyOTA3MDYsIm5iZiI6MTcwNjI5MDQwNiwicGF0aCI6Ii80NTYyNDYxLzMwMDA2MDAzMC1kMjRiYmJhOC03NDNjLTQ4N2EtODRmYy02MjhjNzQ5MTgyMTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjZUMTczMzI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjQ2ZmU0NmQwYmI1N2ZhZDIxNGRjYWY4ZWZhMDJmYjIxY2M4MDMyYjFlZDc5Y2MwZmI3NTIxZTJmOTQyYmZkNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Onma1EoULfGpHRbUtFV26mLnZ91TEaOi27Gfx5D8M5w)
> 
> I'll debug further later tonight.

Disregard, i had to refresh the NPM ui to get the new option showing :)

**Comment by lateautumn233** [comment]
> Hello, I have tried to add HTP3 support for nginx-proxy-manager
> 
> If you are interested in this, you can give it a try
> 
> docker: https://hub.docker.com/repository/docker/lateautumn233/nginx-proxy-manager/ Source code: https://github.com/lateautumn233/nginx-proxy-manager

Other architectures
`docker push lateautumn233/nginx-proxy-manager:test`

I need more testing, and if there are no issues, I will try to contribute upstream

**Comment by RainmakerRaw** [comment]
> Other architectures `docker pull lateautumn233/nginx-proxy-manager:test`
> 
> I need more testing, and if there are no issues, I will try to contribute upstream

Working a-OK on arm64 (Radxa Rock 5 model B), for me. Page loads confirm http/3 and it's working perfectly. I can finally use NPM again!

**Comment by timnolte** [comment]
FYI, if you haven't seen recently that there is at least 1 open CVE for HTTP/3 + QUIC in Nginx. It was actually the subject of one of the core devs leaving the project and forking Nginx. I'd be careful about pushing this forward as the support is still marked experimental.

**Comment by timnolte** [comment]
Here are the details for reference. https://my.f5.com/manage/s/article/K000138444

**Comment by timnolte** [comment]
It does look like a fixed version was released, however there is a part of me that is concerned that the HTTP/3 + QUIC module is considered experimental still.

**Comment by lateautumn233** [comment]
> FYI, if you haven't seen recently that there is at least 1 open CVE for HTTP/3 + QUIC in Nginx. It was actually the subject of one of the core devs leaving the project and forking Nginx. I'd be careful about pushing this forward as the support is still marked experimental.

Perhaps we should switch to free nginx?

**Comment by lastsamurai26** [comment]
Portainer didn't show that utp port 443 or 80 is open ? 
![image](https://github.com/NginxProxyManager/nginx-proxy-manager/assets/3257844/dd6943e7-ebcb-400f-b0ef-8f10c60fd193)

**Comment by tgarif** [comment]
Hi @lateautumn233, I tried your version of NPM + quic enabled but somehow after inspecting the network tab it seems like the protocol doesn't use http3 but instead it uses http/1.1 or http2 (if enabled). I did expose port 443/udp from the docker compose already.

Another thing that I've tried is to add `reuseport` at the listen directive for quic and it does work if I have only one proxy host. If I add a new proxy host, it will not work with `reuseport`. Any advise you can give on my situation?

**Comment by lateautumn233** [comment]
> Hi @lateautumn233, I tried your version of NPM + quic enabled but somehow after inspecting the network tab it seems like the protocol doesn't use http3 but instead it uses http/1.1 or http2 (if enabled). I did expose port 443/udp from the docker compose already.
> 
> Another thing that I've tried is to add `reuseport` at the listen directive for quic and it does work if I have only one proxy host. If I add a new proxy host, it will not work with `reuseport`. Any advise you can give on my situation?

Maybe it's related to this
https://stackoverflow.com/questions/76348128/enabling-quic-http-3-on-multiple-domains-with-nginx-1-25

**Comment by vincent1890** [comment]
update pls ?

**Comment by stanthewizzard** [comment]
On my side I siwtched to caddy. Conf file is finally easier. Http3 out of the box

**Comment by iwyz0** [comment]
http3 is actually very important, I don't understand why it isn't given priority.

**Comment by vincent1890** [comment]
Any update on this? @jc21

**Comment by hersche** [comment]
hi [lateautumn233](https://github.com/lateautumn233), i would like to ask if you can update your image on docker-hub. this project added a acme-plugin recently which i use and which does not allow me to use your version.

hopefully that http3-change is gonna be integrated here soon. thanks a lot for your work!

**Comment by dpalic** [comment]
any update on this issue? now 3.5y old

**Comment by thefeli73** [comment]
Moving to Caddy. NPM is dead.

**Comment by sleighton** [comment]
Guys, no need to announce.  Just go.

**Comment by taimadoCE** [comment]
Nothing about it? Is it planned?

**Comment by thefeli73** [comment]
> Nothing about it? Is it planned?

@taimadoCE check out npm+ (fork) https://github.com/ZoeyVid/NPMplus?tab=readme-ov-file#compatibility-to-upstream

**Comment by taimadoCE** [comment]
> > Nothing about it? Is it planned?
> 
> [@taimadoCE](https://github.com/taimadoCE) check out npm+ (fork) https://github.com/ZoeyVid/NPMplus?tab=readme-ov-file#compatibility-to-upstream

I'm in linux/arm/v7. That fork doesn't support it.

**Comment by LDprg** [comment]
@taimadoCE use caddy instead, it has http3 support . You won't get a gui, but I switched some time ago and I am very happy with it.

**Comment by taimadoCE** [comment]
> [@taimadoCE](https://github.com/taimadoCE) use caddy instead, it has http3 support . You won't get a gui, but I switched some time ago and I am very happy with it.

Yes...I need to investigate how to migrate my local domains with duckDNS to get http3 support on them...

Example: portainer.myname.duckdns.org

**Comment by XpycTee** [comment]
> Issue is now considered stale. If you want to keep it open, please comment 👍

👍
